File: apps/mobile/android/app/src/main/java/com/aurelia/app/auth/AuthInterceptor.kt
package com.aurelia.app.auth

import android.util.Log

object AuthInterceptor {
    private const val TAG = "AuthInterceptor"
    private var logoutCallback: (() -> Unit)? = null

    fun setLogoutCallback(callback: () -> Unit) {
        logoutCallback = callback
    }

    fun clearLogoutCallback() {
        logoutCallback = null
    }

    fun isUnauthorizedError(error: Throwable): Boolean {
        val message = error.message?.lowercase() ?: return false
        return message.contains("unauthorized") ||
            message.contains("401") ||
            message.contains("authentication") ||
            message.contains("not authenticated")
    }

    fun isUnauthorizedError(errorMessage: String?): Boolean {
        val message = errorMessage?.lowercase() ?: return false
        return message.contains("unauthorized") ||
            message.contains("401") ||
            message.contains("authentication") ||
            message.contains("not authenticated")
    }

    fun handlePotentialAuthError(error: Throwable): Boolean {
        if (isUnauthorizedError(error)) {
            Log.w(TAG, "Unauthorized error detected, triggering logout")
            triggerLogout()
            return true
        }
        return false
    }

    fun handlePotentialAuthError(errorMessage: String?): Boolean {
        if (isUnauthorizedError(errorMessage)) {
            Log.w(TAG, "Unauthorized error detected, triggering logout")
            triggerLogout()
            return true
        }
        return false
    }

    private fun triggerLogout() {
        logoutCallback?.invoke() ?: Log.w(TAG, "Logout callback not set")
    }
}

File: apps/mobile/ios/Aurelia/Services/AuthInterceptor.swift
import Foundation
import os

/// Detects 401/unauthorized errors from Rust API calls and triggers automatic logout.
final class AuthInterceptor: @unchecked Sendable {
    static let shared = AuthInterceptor()

    private let logger = Logger(subsystem: "com.aurelia.app", category: "AuthInterceptor")
    private var logoutCallback: (() -> Void)?

    private init() {}

    func setLogoutCallback(_ callback: @escaping () -> Void) {
        logoutCallback = callback
    }

    func clearLogoutCallback() {
        logoutCallback = nil
    }

    func isUnauthorizedError(_ error: Error) -> Bool {
        isUnauthorizedError(error.localizedDescription)
    }

    func isUnauthorizedError(_ message: String?) -> Bool {
        guard let message = message?.lowercased() else { return false }
        return message.contains("unauthorized")
            || message.contains("401")
            || message.contains("authentication")
            || message.contains("not authenticated")
    }

    /// Returns true if the error was an auth error and logout was triggered.
    @discardableResult
    func handlePotentialAuthError(_ error: Error) -> Bool {
        handlePotentialAuthError(error.localizedDescription)
    }

    @discardableResult
    func handlePotentialAuthError(_ message: String?) -> Bool {
        if isUnauthorizedError(message) {
            logger.warning("Unauthorized error detected, triggering logout")
            logoutCallback?()
            return true
        }
        return false
    }
}
